lecture-notes/3-bld.md

```diff
@@ -418,20 +420,284 @@ order:
+x := $(patsubst bld/%.o,src/%.c,$(objs))
+# foo.c bar.c
+srcs := foo.c bar.c bar.h
+x := $(filter %.c,$(srcs))
+```
+
+You'll usually use functions when specifying variables, targets, or
+prerequisites: inside a recipe itself, it's almost always better to use the
+shell to do any text processing you need. <!-- TODO: add a footnote explaining
+why -->
+
+Function invocations take one or more comma-separated arguments, which is what
+distinguishes them from variable expansions. They otherwise have identical
+`$()` syntax.
+
+Functions are are (in general) evaluated the same way as variables are
+expanded---at the time they're used in a rule or a `:=` assignment.
+
 ### Make variables and functions are textual
 
+From reading about variable expansions and function expansions and the various
+phases of everything, you may have gotten the feeling that, similar to the C
+preprocessor and shell, everything is textual. You would be correct.
+
+Like the shell, Make has no concept of data types. Everything is text, and
+Make's minimal escaping and tolerance for spaces can lead to some pretty odd
+constructs. The following is taken verbatim from the GNU Make manual:
+
+```make
+comma:= ,
+empty:=
+space:= $(empty) $(empty)
+foo:= a b c
+bar:= $(subst $(space),$(comma),$(foo))
+# bar is now 'a,b,c'.
+```
+
+We will leave you to marinate with that one.
+
 ## Lecture 3
 
+You've learned about variables and some built-in functions. With just that, you
+can write quite a reasonable Makefile for a little project. But you will soon
+find yourself wanting features like the following.
+
 ### Pattern rules
+
+Similar to the patterns you saw in `patsubst`, rules can have patterns. To
+specify a pattern rule, use `%` somewhere in the target name. For example:
+
+```make
+%.o: %.c %.h
+	gcc -c $< -o $@
+```
+
+In this rule, the `%` matches any (non-empty) text before `.o` and is called
+the *stem*. This means it will match any target like `a.o`, `foo.o`, etc. When
+it matches, Make substitutes the stem (`a` or `foo`) in the `%` in the
+prerequisites as well.
+
+<!-- TODO(tom): clarify whether multi-target pattern rules are different from
+normal "multi-target" rules. they seem different?? -->
+
+In these pattern rules, automatic variables do not only remain convenient; they
+are also a necessity.
+
 ### Implicit rules
+
+If you have ever had `foo.c` sitting around and no Makefile yet, but
+accidentally run `make`, you may have noticed Make... still do something? This
+is because of *implicit rules*. These "implicit rules" are automatically
+present any time you run `make` even if you don't have a Makefile at all!
+
+Make already knows how to compile C, C++, Fortran, and many other languages:
+inside Make are built-in pattern rules for common file extensions (e.g. `.c`,
+`.cpp`, `.o`).
+
+Because they are built-in, they support limited customization. If the implicit
+rule makes use of variables (such as `CC`, `CFLAGS`, and `LDFLAGS`), then you
```

Comment:
Talk about how to figure out if an implicit rule uses variables? I generally just hope the Make docs are correct/complete, which they usually are but sometimes aren't. But perhaps there's an easy-to-understand source of truth that would be a better reference?

Reply:
demo reading docs and reading source to verify; that's the point of the class